cpp/src/strings/search/find.cu

```diff
@@ -414,6 +418,289 @@ std::unique_ptr<column> contains_warp_parallel(strings_column_view const& input,
+    d_results[target_idx][str_idx] = d_target.size_bytes() == 0;
+  }
+
+  for (auto str_byte_idx = 0; str_byte_idx < d_str.size_bytes();
+       ++str_byte_idx) {  // iterate the start index in the string
+
+    // binary search in the target first char set.
+    char c = *(d_str.data() + str_byte_idx);
+    auto first_byte_ptr =
+      thrust::lower_bound(thrust::seq, d_target_first_bytes.begin(), d_target_first_bytes.end(), c);
+
+    if (not(first_byte_ptr != d_target_first_bytes.end() && *first_byte_ptr == c)) {
+      // For non-empty targets: no need to search for `str_byte_idx` position, because first char is
+      // unmatched. For empty targets: already set result as found.
+      continue;
+    }
+
+    int first_char_index_in_list = first_byte_ptr - d_target_first_bytes.begin();
+    // get possible targets
+    auto const possible_targets_list =
+      cudf::list_device_view{d_target_indexes_for_first_bytes, first_char_index_in_list};
+
+    for (auto i = 0; i < possible_targets_list.size(); ++i) {  // iterate possible targets
+      auto target_idx = possible_targets_list.element<size_type>(i);
+      if (!d_results[target_idx][str_idx]) {  // not found before
+        auto const d_target = d_targets.element<string_view>(target_idx);
+        if (d_str.size_bytes() - str_byte_idx >= d_target.size_bytes()) {
+          // first char already checked, only need to check the [2nd, end) chars if has.
+          bool found = true;
+          for (auto i = 1; i < d_target.size_bytes(); i++) {
+            if (*(d_str.data() + str_byte_idx + i) != *(d_target.data() + i)) {
+              found = false;
+              break;
+            }
+          }
+          if (found) { d_results[target_idx][str_idx] = true; }
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Execute multi contains for short strings
+ * First index the first char for all targets.
+ * Index the first char:
+ *   collect first char for all targets and do uniq and sort,
+ *   then index the targets for the first char.
+ *   e.g.:
+ *     targets: xa xb ac ad af
+ *     first char set is: (a, x)
+ *     index result is:
+ *       {
+ *         a: [2, 3, 4],   // indexes for: ac ad af
+ *         x: [0, 1]       // indexes for: xa xb
+ *       }
+ * when do searching:
+ *   find (binary search) from `first char set` for a char in string:
+ *     if char in string is not in ['a', 'x'], fast skip
+ *     if char in string is 'x', then only need to try ["xa", "xb"] targets.
+ *     if char in string is 'a', then only need to try ["ac", "ad", "af"] targets.
+ *
+ */
+std::vector<std::unique_ptr<column>> multi_contains_using_indexes(
+  strings_column_view const& input,
+  strings_column_view const& targets,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  auto const num_targets = static_cast<size_type>(targets.size());
+  CUDF_EXPECTS(not targets.is_empty(), "Must specify at least one target string.");
+
+  // 1. copy targets from device to host
```

Comment:
This seems complicated. I'm not sure copying to the host is necessary. The targets are in a strings column and so the offsets of the column already point to the first character. Also you could use the [`cudf::sorted_order`](https://docs.rapids.ai/api/libcudf/stable/group__column__sort#ga0a8af4fc5e6350716261e6940c0563eb) on the targets to get the indices of the those offsets in sorted order.

Reply:
Normally, the num of targets is not large, so could we keep the current way?
IIUC, sorting, uniq and collect the indexes on CPU is fast than on GPU.
If we use GPU, we need to invokes several kernels to do sort/uniq/collect the indexes.

Reply:
I don't expect it will require many kernels and I feel the kernels would replace the host-device copies and so may run faster.

Reply:
If using GPU, my code will like:

1. make a mumeric column to save target index, values are: [0, 1, 2, 3, ... target_size)
create int column with target_size elements.
fill values:  thrust::sequence(target_indexes.begin(), target_indexes.end())

2. filter out empty target
input: targets, target_indexes
output: targets_after_filter, target_indexes_after_filter

3. make a column to save first target char
first_char
get a table (first_char, target_indexes_after_filter)

4. group by first char and collect indexes
Input: table (first_char, target_indexes_after_filter)
output: index list column

5. concatenate first chars into a string_view which is used for binary search.

Above steps are like: project -> filter -> agg
@davidwendt  do you have a simple solution?

Reply:
Yes, I think so. I can work on it next week.

Reply:
I have a solution that is a bit simpler (less code) and runs about the same or faster for most cases.
I can send you the `find.cu` file to look at.
I'm not able to get the warp-parallel kernel in this PR to work. I can provide the test case I was using for this.

Reply:
I checked your way to build indexes:
```cpp
  // build sorted-order indices and a vector of first-bytes of the targets
  auto indices = cudf::sorted_order(table_view({targets.parent()}));
  rmm::device_uvector<u_char> first_bytes(targets.size(), stream);
  thrust::transform(rmm::exec_policy(stream),
                    indices->view().begin<size_type>(),
                    indices->view().end<size_type>(),
                    first_bytes.begin(),
                    [d_targets = *d_targets] __device__(size_type idx) -> u_char {
                      auto d_tgt = d_targets.element<string_view>(idx);
                      return d_tgt.empty() ? u_char{0} : static_cast<u_char>(d_tgt.data()[0]);
                    });
```
More steps are needed to deduplicate first_bytes, filter empty targets.

I tested the current way, in most cases, it takes less than 1ms:
```
Used time is 0.000036 seconds.
Used time is 0.000027 seconds.
Used time is 0.000028 seconds.
Used time is 0.000476 seconds.
Used time is 0.001648 seconds.
Used time is 0.000026 seconds.
Used time is 0.000027 seconds.
Used time is 0.000027 seconds.
Used time is 0.000026 seconds.
Used time is 0.000028 seconds.
Used time is 0.000025 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000027 seconds.
Used time is 0.000025 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000027 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000027 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000026 seconds.
Used time is 0.000027 seconds.
Used time is 0.000029 seconds.
Used time is 0.000030 seconds.
Used time is 0.000029 seconds.
Used time is 0.000030 seconds.
Used time is 0.001938 seconds.
Used time is 0.000037 seconds.
Used time is 0.000030 seconds.
Used time is 0.000030 seconds.
Used time is 0.000029 seconds.
Used time is 0.000029 seconds.
Used time is 0.000029 seconds.
Used time is 0.000039 seconds.
Used time is 0.000035 seconds.
Used time is 0.000029 seconds.
Used time is 0.000029 seconds.
Used time is 0.000030 seconds.
Used time is 0.000029 seconds.
Used time is 0.000034 seconds.
Used time is 0.000027 seconds.
```

Reply:
I don't think the deduplicate is necessary since `lower_bound` is very fast either way.

Reply:
After this [commit](https://github.com/rapidsai/cudf/pull/16641/commits/6741bef90237ae4c0e34eec13e96bb15c66ce38f)
Indexing time of 16 strings with each string contains 100 chars:
index used time: 0.000128523 seconds.  Less than 1ms.

Reply:
If most the first chars are identical, then it's reasonable to remove deduplicating chars.

Reply:
If most the first chars are identical, then lower-bound will still not be an issue compared to the for-loop for possible-targets. My experiments show that lower-bound handles duplicates quickly enough and not worth the overhead of removing them.

Reply:
Here is the benchmark results with simpler code that does not require copying data to the host or removing duplicates:
```
| row_width | num_rows |hit_rate|   Ref Time |   Cmp Time |        Diff |   %Diff |
|-----------|----------|--------|------------|------------|-------------|---------|
|    32     |  260000  |   20   | 302.378 us | 288.183 us |  -14.195 us |  -4.69% |
|    64     |  260000  |   20   | 504.029 us | 480.324 us |  -23.705 us |  -4.70% |
|    128    |  260000  |   20   | 167.120 us | 151.788 us |  -15.332 us |  -9.17% |
|    256    |  260000  |   20   | 193.492 us | 173.841 us |  -19.651 us | -10.16% |
|    512    |  260000  |   20   | 227.784 us | 219.248 us |   -8.536 us |  -3.75% |
|   1024    |  260000  |   20   | 320.813 us | 312.556 us |   -8.257 us |  -2.57% |
|    32     | 1953000  |   20   |   1.394 ms |   1.383 ms |  -10.229 us |  -0.73% |
|    64     | 1953000  |   20   |   2.645 ms |   2.622 ms |  -22.526 us |  -0.85% |
|    128    | 1953000  |   20   | 477.524 us | 463.520 us |  -14.004 us |  -2.93% |
|    256    | 1953000  |   20   | 609.522 us | 600.221 us |   -9.300 us |  -1.53% |
|    512    | 1953000  |   20   | 889.037 us | 883.210 us |   -5.828 us |  -0.66% |
|   1024    | 1953000  |   20   |   1.459 ms |   1.459 ms |   -0.025 us |  -0.00% |
|    32     | 16777216 |   20   |  11.162 ms |  11.185 ms |   23.249 us |   0.21% |
|    64     | 16777216 |   20   |  21.364 ms |  21.341 ms |  -23.421 us |  -0.11% |
|    32     |  260000  |   80   | 245.660 us | 232.903 us |  -12.758 us |  -5.19% |
|    64     |  260000  |   80   | 468.730 us | 414.366 us |  -54.364 us | -11.60% |
|    128    |  260000  |   80   | 147.905 us | 139.493 us |   -8.412 us |  -5.69% |
|    256    |  260000  |   80   | 169.207 us | 161.092 us |   -8.115 us |  -4.80% |
|    512    |  260000  |   80   | 214.916 us | 206.459 us |   -8.458 us |  -3.94% |
|   1024    |  260000  |   80   | 307.762 us | 299.603 us |   -8.159 us |  -2.65% |
|    32     | 1953000  |   80   |   1.196 ms |   1.173 ms |  -23.318 us |  -1.95% |
|    64     | 1953000  |   80   |   2.251 ms |   2.169 ms |  -81.940 us |  -3.64% |
|    128    | 1953000  |   80   | 420.562 us | 411.235 us |   -9.327 us |  -2.22% |
|    256    | 1953000  |   80   | 568.573 us | 562.310 us |   -6.263 us |  -1.10% |
|    512    | 1953000  |   80   | 868.699 us | 867.512 us |   -1.187 us |  -0.14% |
|   1024    | 1953000  |   80   |   1.479 ms |   1.486 ms |    6.963 us |   0.47% |
|    32     | 16777216 |   80   |   9.472 ms |   9.369 ms | -102.513 us |  -1.08% |
|    64     | 16777216 |   80   |  17.781 ms |  17.407 ms | -374.020 us |  -2.10% |
```

Reply:
Thanks for testing this.
I have an alternate solution that looks to be up to 3x faster than this without copying any data to the host.
Perhaps I should push an alternate PR to more easily share code.

Reply:
So the 3x faster seemed to be fluke and likely a bug in my code that has been fixed now. 
I've created #16900 as an alternate solution based on this one.
I've refactored the code quite a bit and simplified some of the logic.
The performance is about the same for the most part.
I removed the target chunking and opted for global memory when the shared-memory threshold was reached.
It may be possible to use more shared-memory in the warp-parallel case. There is probably still more to do in this area I think.